### ISSUE TITLE
Improve API for setting Configurations as producers or consumers

### ISSUE THREAD

**Issue by aSemy** [gradle/gradle#21701]
I find it very difficult to remember how to configure Configurations as producers and consumers The existing properties, `isCanBeResolved` and `isCanBeConsumed` are vague (they seem like synonyms to me). The documentation can help - particularly [the table in this page](https://docs.gradle.org/7.5.1/userguide/declaring_dependencies.html#sec:resolvable-consumable-configs) - but finding it is always difficult because it doesn't have a HTML link and it's buried in the middle of a large page.

I'd like the API to be changed so it's more descriptive and clear, and the Javadoc updated. 

I'd also like the old example for JaCoCo to be updated and re-instated https://docs.gradle.org/7.3/samples/sample_jvm_multi_project_with_code_coverage.html. It was removed because of the new plugin, but knowing how to share artifacts between plugins is still useful.

### Expected Behavior

What I think would be best is to do away with the multiple booleans, and replace them with an enum. Under the hood, there can still be the boolean flags.

```kotlin
/**
 * <insert more docs here>
 * 
 * See: https://docs.gradle.org/7.5.1/userguide/declaring_dependencies.html#sec:resolvable-consumable-configs
 */
enum class ConfigurationRole(
    val isCanBeConsumed: Boolean,
    val isCanBeResolved: Boolean,
) {
    /** Bucket of dependencies */
    INTERNAL(false, false),
    /** Exposed to consumers */
    PRODUCER(true, false),
    /** Resolve for certain usage */
    INCOMING(false, true),
    /** Legacy, don’t use */
    LEGACY(true, true),
    ;
}
```

In the Kotlin DSL this is easy to apply already, with an extension var.

```kotlin
val jacocoAggregateSource: Configuration by configurations.creating {
  role = ConfigurationRole.PRODUCER
}

var Configuration.role: ConfigurationRole
    get() = ConfigurationRole.get(isCanBeResolved, isCanBeConsumed) // helper function
    set(value) {
        isCanBeConsumed = value.isCanBeConsumed
        isCanBeResolved = value.isCanBeResolved
    }
```

Another option is to create some helper methods. On that helper method, I'd like there to be descriptive Javadoc, and a direct link to the pertinent documentation.

```kotlin
val jacocoAggregateSource: Configuration by configurations.creating {
  isProducer() // new helper method
}

/**
 * Mark this [Configuration] as a 'producer' that exposes artifacts and their dependencies for consumption by other
 * projects
 *
 * See: https://docs.gradle.org/7.5.1/userguide/declaring_dependencies.html#sec:resolvable-consumable-configs
 */
fun Configuration.isProducer() {
    isCanBeResolved = false
    isCanBeConsumed = true
}
```

### Current Behavior

It's difficult to create configurations with the correct flag. It's difficult to find the documentation. The words 'resolvable' and 'consumable' seem like synonyms, I don't really get the difference between them.

### Context

I'm trying again to share outputs between projects in a safe way.

**Comment by jbartok** [comment]
Yes, this is something we want to do, but it's quite a large change and not actually planned for the short term.
